File: services/api/dto/daily_reports.go
package dto

import "time"

// ── Request types ─────────────────────────────────────────────────────────────

type WorkerHoursInput struct {
	WorkerID    string  `json:"worker_id" binding:"required"`
	HoursWorked float64 `json:"hours_worked"`
	Notes       *string `json:"notes"`
}

type ActivityInput struct {
	ProjectMaterialID  *string `json:"project_material_id"`
	CustomMaterialName *string `json:"custom_material_name"`
	Quantity           float64 `json:"quantity"`
	Unit               string  `json:"unit"`
	ActivityType       string  `json:"activity_type"`
	IsVTK              bool    `json:"is_vtk"`
	Notes              *string `json:"notes"`
}

type CreateDailyReportRequest struct {
	ProjectID   string             `json:"project_id" binding:"required"`
	ReportDate  string             `json:"report_date" binding:"required"`
	Notes       *string            `json:"notes"`
	WorkerHours []WorkerHoursInput `json:"worker_hours"`
	Activities  []ActivityInput    `json:"activities"`
}

type RejectDailyReportRequest struct {
	Reason string `json:"reason"`
}

// ── List response ─────────────────────────────────────────────────────────────

type DailyReportListItem struct {
	ID               string     `json:"id"`
	ProjectID        string     `json:"project_id"`
	ProjectName      string     `json:"project_name"`
	PoslovodaID      string     `json:"poslovoda_id"`
	PoslovodaName    string     `json:"poslovoda_name"`
	ReportDate       string     `json:"report_date"`
	Status           string     `json:"status"`
	WorkerHoursCount int        `json:"worker_hours_count"`
	TotalHours       float64    `json:"total_hours"`
	ActivitiesCount  int        `json:"activities_count"`
	SubmittedAt      *time.Time `json:"submitted_at"`
	CreatedAt        time.Time  `json:"created_at"`
	UpdatedAt        time.Time  `json:"updated_at"`
}

// ── Detail response ───────────────────────────────────────────────────────────

type DailyReportProject struct {
	ID     string `json:"id"`
	Name   string `json:"name"`
	Status string `json:"status"`
}

type DailyReportPerson struct {
	ID        string `json:"id"`
	FirstName string `json:"first_name"`
	LastName  string `json:"last_name"`
	FullName  string `json:"full_name"`
}

type DailyReportWorkerHours struct {
	ID          string  `json:"id"`
	WorkerID    string  `json:"worker_id"`
	WorkerName  string  `json:"worker_name"`
	HoursWorked float64 `json:"hours_worked"`
	Notes       *string `json:"notes"`
}

type DailyReportActivity struct {
	ID                 string  `json:"id"`
	ProjectMaterialID  *string `json:"project_material_id"`
	MaterialName       string  `json:"material_name"`
	CustomMaterialName *string `json:"custom_material_name"`
	Quantity           float64 `json:"quantity"`
	Unit               string  `json:"unit"`
	ActivityType       string  `json:"activity_type"`
	IsVTK              bool    `json:"is_vtk"`
	Notes              *string `json:"notes"`
}

type DailyReportAttachment struct {
	ID           string    `json:"id"`
	OriginalName string    `json:"original_name"`
	ContentType  string    `json:"content_type"`
	FileSize     int64     `json:"file_size"`
	CreatedAt    time.Time `json:"created_at"`
}

type DailyReportDetail struct {
	ID          string                   `json:"id"`
	Project     DailyReportProject       `json:"project"`
	Poslovoda   DailyReportPerson        `json:"poslovoda"`
	ReportDate  string                   `json:"report_date"`
	Status      string                   `json:"status"`
	Notes       *string                  `json:"notes"`
	SubmittedAt *time.Time               `json:"submitted_at"`
	CreatedAt   time.Time                `json:"created_at"`
	UpdatedAt   time.Time                `json:"updated_at"`
	WorkerHours []DailyReportWorkerHours `json:"worker_hours"`
	Activities  []DailyReportActivity    `json:"activities"`
	Attachments []DailyReportAttachment  `json:"attachments"`
}

// ── Form-data response ────────────────────────────────────────────────────────

type FormDataProject struct {
	ID   string `json:"id"`
	Name string `json:"name"`
}

type FormDataWorker struct {
	ID        string `json:"id"`
	FirstName string `json:"first_name"`
	LastName  string `json:"last_name"`
	FullName  string `json:"full_name"`
}

type FormDataMaterial struct {
	ID                string  `json:"id"`
	MaterialName      string  `json:"material_name"`
	MaterialCode      *string `json:"material_code"`
	AvailableQuantity float64 `json:"available_quantity"`
	Unit              string  `json:"unit"`
}

type DailyReportFormData struct {
	Projects  []FormDataProject  `json:"projects"`
	Workers   []FormDataWorker   `json:"workers"`
	Materials []FormDataMaterial `json:"materials"`
}

// ── Filter ────────────────────────────────────────────────────────────────────

type DailyReportFilter struct {
	ProjectID   *string
	PoslovodaID *string // scopes to a specific poslovoda (used for poslovoda role)
	DateFrom    *string
	DateTo      *string
	Status      *string
}

File: apps/web/lib/types/daily-reports.ts
// ── Enums / labels ────────────────────────────────────────────────────────────

export type ReportStatus = 'draft' | 'submitted' | 'approved' | 'rejected'
export type ActivityType = 'montaza' | 'demontaza' | 'other'

export const REPORT_STATUS_LABELS: Record<ReportStatus, string> = {
  draft: 'Nacrt',
  submitted: 'Predano',
  approved: 'Odobreno',
  rejected: 'Odbijeno',
}

export const ACTIVITY_TYPE_LABELS: Record<ActivityType, string> = {
  montaza: 'Montaža',
  demontaza: 'Demontaža',
  other: 'Ostalo',
}

// ── API response types ────────────────────────────────────────────────────────

export interface DailyReportListItem {
  id: string
  project_id: string
  project_name: string
  poslovoda_id: string
  poslovoda_name: string
  report_date: string
  status: ReportStatus
  worker_hours_count: number
  total_hours: number
  activities_count: number
  submitted_at: string | null
  created_at: string
  updated_at: string
}

export interface DailyReportProject {
  id: string
  name: string
  status: string
}

export interface DailyReportPerson {
  id: string
  first_name: string
  last_name: string
  full_name: string
}

export interface DailyReportWorkerHours {
  id: string
  worker_id: string
  worker_name: string
  hours_worked: number
  notes: string | null
}

export interface DailyReportActivity {
  id: string
  project_material_id: string | null
  material_name: string
  custom_material_name: string | null
  quantity: number
  unit: string
  activity_type: ActivityType
  is_vtk: boolean
  notes: string | null
  tracking_type: 'stock' | 'work'
}

export interface DailyReportAttachment {
  id: string
  original_name: string
  content_type: string
  file_size: number
  created_at: string
}

export interface DailyReportDetail {
  id: string
  project: DailyReportProject
  poslovoda: DailyReportPerson
  report_date: string
  status: ReportStatus
  notes: string | null
  submitted_at: string | null
  created_at: string
  updated_at: string
  worker_hours: DailyReportWorkerHours[]
  activities: DailyReportActivity[]
  attachments: DailyReportAttachment[]
}

// ── Form data types ───────────────────────────────────────────────────────────

export interface FormDataProject {
  id: string
  name: string
}

export interface FormDataWorker {
  id: string
  first_name: string
  last_name: string
  full_name: string
}

export interface FormDataMaterial {
  id: string
  material_name: string
  material_code: string | null
  available_quantity: number
  unit: string
  tracking_type: 'stock' | 'work'
}

export interface DailyReportFormData {
  projects: FormDataProject[]
  workers: FormDataWorker[]
  materials: FormDataMaterial[]
}

// ── Request payload types ─────────────────────────────────────────────────────

export interface WorkerHoursInput {
  worker_id: string
  hours_worked: number
  notes: string | null
}

export interface ActivityInput {
  project_material_id: string | null
  custom_material_name: string | null
  quantity: number
  unit: string
  activity_type: ActivityType
  is_vtk: boolean
  notes: string | null
  tracking_type?: 'stock' | 'work'
}

export interface CreateDailyReportPayload {
  project_id: string
  report_date: string
  notes: string | null
  worker_hours: WorkerHoursInput[]
  activities: ActivityInput[]
}

// ── UI-local types (used inside form components) ──────────────────────────────

/** Per-worker input row inside WorkerHoursEditor */
export interface WorkerHoursEntry {
  worker_id: string
  worker_name: string
  hours: string   // raw string input
  notes: string
}

/** Activity row inside the preview list (has a temp id for keying) */
export interface ActivityInputUI extends ActivityInput {
  _tempId: string
  _displayName: string  // resolved material name for display
  tracking_type?: 'stock' | 'work'
}

// ── Permission helpers ────────────────────────────────────────────────────────

export function canCreateDailyReport(role: string): boolean {
  return role === 'poslovoda' || role === 'direktor' || role === 'inzenjer'
}

export function canEditDailyReport(role: string, callerEmpId: string | null, poslovodaId: string, status: ReportStatus): boolean {
  if (status === 'approved') return false
  if (role === 'direktor' || role === 'inzenjer') return true
  if (role === 'poslovoda') return callerEmpId === poslovodaId
  return false
}

export function canApproveReject(role: string): boolean {
  return role === 'direktor' || role === 'inzenjer'
}
